File: Leetcode/0082-remove-duplicates-from-sorted-list-ii/solution.go
package main

import "fmt"

// 0082. Remove Duplicates from Sorted List II
// Time:  O(n), Space: O(1)

type ListNode struct {
	Val  int
	Next *ListNode
}

func deleteDuplicates(head *ListNode) *ListNode {
	dummy := &ListNode{Next: head}
	prev := dummy
	cur := head
	for cur != nil {
		if cur.Next != nil && cur.Next.Val == cur.Val {
			v := cur.Val
			for cur != nil && cur.Val == v {
				cur = cur.Next
			}
			prev.Next = cur
		} else {
			prev = cur
			cur = cur.Next
		}
	}
	return dummy.Next
}

// helpers
func toList(arr []int) *ListNode {
	if len(arr) == 0 {
		return nil
	}
	head := &ListNode{Val: arr[0]}
	cur := head
	for _, v := range arr[1:] {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return head
}

func toSlice(head *ListNode) []int {
	out := []int{}
	for head != nil {
		out = append(out, head.Val)
		head = head.Next
	}
	return out
}

func eq(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func main() {
	passed, failed := 0, 0
	test := func(name string, got, expected []int) {
		if eq(got, expected) {
			fmt.Printf("PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("FAIL: %s got=%v\n", name, got)
			failed++
		}
	}

	cases := []struct {
		name string
		in   []int
		want []int
	}{
		{"Example 1", []int{1, 2, 3, 3, 4, 4, 5}, []int{1, 2, 5}},
		{"Example 2", []int{1, 1, 1, 2, 3}, []int{2, 3}},
		{"Empty", []int{}, []int{}},
		{"All duplicates", []int{1, 1, 1, 1}, []int{}},
		{"No duplicates", []int{1, 2, 3, 4}, []int{1, 2, 3, 4}},
		{"Single", []int{5}, []int{5}},
		{"Duplicate tail", []int{1, 2, 3, 3}, []int{1, 2}},
		{"Multiple runs", []int{1, 1, 2, 3, 3, 4, 5, 5}, []int{2, 4}},
		{"Negatives", []int{-2, -1, -1, 0, 1}, []int{-2, 0, 1}},
	}
	for _, c := range cases {
		test(c.name, toSlice(deleteDuplicates(toList(c.in))), c.want)
	}
	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}

File: Leetcode/0082-remove-duplicates-from-sorted-list-ii/solution.py
from typing import Optional, List


class ListNode:
    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next


class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """Time O(n), Space O(1)."""
        dummy = ListNode(0, head)
        prev = dummy
        cur = head
        while cur:
            if cur.next and cur.next.val == cur.val:
                v = cur.val
                while cur and cur.val == v:
                    cur = cur.next
                prev.next = cur
            else:
                prev = cur
                cur = cur.next
        return dummy.next


def to_list(arr):
    if not arr: return None
    head = ListNode(arr[0])
    cur = head
    for v in arr[1:]:
        cur.next = ListNode(v)
        cur = cur.next
    return head


def to_arr(h):
    out = []
    while h:
        out.append(h.val)
        h = h.next
    return out


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    def test(name, got, expected):
        global passed, failed
        if got == expected: print(f"PASS: {name}"); passed += 1
        else: print(f"FAIL: {name} got={got}"); failed += 1

    cases = [
        ("Example 1", [1, 2, 3, 3, 4, 4, 5], [1, 2, 5]),
        ("Example 2", [1, 1, 1, 2, 3], [2, 3]),
        ("Empty", [], []),
        ("All duplicates", [1, 1, 1, 1], []),
        ("No duplicates", [1, 2, 3, 4], [1, 2, 3, 4]),
        ("Single", [5], [5]),
        ("Duplicate tail", [1, 2, 3, 3], [1, 2]),
        ("Multiple runs", [1, 1, 2, 3, 3, 4, 5, 5], [2, 4]),
        ("Negatives", [-2, -1, -1, 0, 1], [-2, 0, 1]),
    ]
    for n, a, exp in cases: test(n, to_arr(sol.deleteDuplicates(to_list(a))), exp)
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
